### ISSUE TITLE
Continuous saving of history

### ISSUE THREAD

**Issue by AckslD** [isamert/scli#112]
If I understand it correctly history is only saved on exit. Would it also be possible to save continuously from time to time? Since sometimes if I just shutdown my computer without explicitly closing scli it seems the history is not saved.

**Comment by exquo** [comment]
Yes, history is currently saved on exit. Being able to protect it from OS crashes, power failures, etc would be an improvement.

**Comment by NK308** [comment]
I guess, making scli react properly to a SIGTERM would already do a lot. Not about crashed, but about beeing shutdown by the terminating desktop session.

**Comment by obnoxiousish** [comment]
Might pull request this in as a feature in next few days, the code has no comments and I want to make something that works clean, might not though.

**Comment by exquo** [comment]
@NK308, thanks for the suggestion! I've added a handler for the `SIGTERM` and `SIGHUP` signals. It attempts to close and save everything gracefully, like it currently does for `SIGINT` (`Ctrl + C`). 
(The changes are in the `develop` branch)

Yes, this still does not protect the history file from system crashes or hangups. To limit the damage in those events, I've also added instructions to backup the history file on startup. Now the loss is limited to the last session, rather than the whole JSON file getting corrupted.

A proper long-term solution would be to store the conversations history in a database instead of a JSON file. This would be more efficient and reliable, since the new messages could be added incrementally, rather than the whole JSON structure re-written on every exit.

**Comment by exquo** [comment]
@obnoxiousish, the signals handling should be taken care of, AFAICT, but feel free to make a PR with additional changes for this or other issues.

Yes, the code documentation is not extensive. The code itself is somewhat modular - you don't have to understand everything to make improvements - but having an overview of how the pieces fit together helps. Feel free to open a new discussion if you have questions.

**Comment by NK308** [comment]
It seems to work now with the development branch.

Messages now survive, when I log out of i3, or reboot, while scli is still open.

**Comment by NK308** [comment]
I fear, I have to correct my previous statement. Writing messages to myself, between scli and android, leads to a correctly saved chat (which was my way of testing the commit), but chats with other contacts don't seem to be saved, like before.

**Comment by mark2185** [comment]
Are there any plans to make a periodic history save? 

E.g. once a day or something, just so one doesn't have to manually quit and start the chat again?

**Comment by exquo** [comment]
This might be a good stop-gap measure until we implement a proper database.

**Comment by mlncn** [comment]
Typing `:quit` is supposed to be one of the ways that shuts down safely with history saved, yes?  Does something have to be configured to enable history?

I have it running via a git checkout, branch `master`, last commit is 2022-06-03 exquo Update signal-cli's required version in README 6ad260c so history save on quit should be working, correct?  It is not for me!

**Comment by moppman** [comment]
> Does something have to be configured to enable history?
> 
See https://github.com/isamert/scli/blob/master/README.md#L184

You need to put `save-history = true` into your `sclirc` file.

**Comment by meZeex** [comment]
bumping this issue, as it happened to me multiple times the last days

i'm using a battery powered device, which can run out of power and even if scli was just idling (aka no messages) for a long time, the session history is gone

it would be nice to have the history saved, either on every message (ideal) or every X minutes
i guess a config option could make sense (eg: update-history = [message | minutes])
